### PR TITLE
[avro] Use short names whenever possible

### LINKED ISSUE #14333: `mz_avro` encodes messages in the default namespace incorrectly
### What version of Materialize are you using?

Latest main (f029b1410c1c181eea10e3910368a4639bff9853)

### How did you install Materialize?

Built from source

### What is the issue?

When the `mz_avro` crate serializes a schema, and it contains a type whose name is in the default (empty) namespace, the serialized schema will include a leading `.` before the type name. For example, a type named `"MyType"` will be serialized as `".MyType"`.

The logic for encoding type names appears to be here: https://github.com/MaterializeInc/materialize/blob/f029b1410c1c181eea10e3910368a4639bff9853/src/avro/src/schema.rs#L558

[This validation logic](https://github.com/apache/avro/blob/1008ce35bff83f7f98d5bd06b03b25c387efd4fc/lang/py/avro/name.py#L33-L47) (`_validate_fullname()`) in the Avro Python library fails on the serialized schema, because it splits the name on `'.'` and the regex chokes because the first component of the name is empty. Other tools in the Apache ecosystem appear to depend on the exact full-name string, and are sadly not robust to the presence of namespaces.

The easiest way to reproduce this is to add a `println!` in `mz_avro::schema::tests::test_schema_fingerprint()`. The fingerprinted schema has a name of `".test"`:

```
{"type":"record","name":".test","fields":[{"name":"a","type":"long","default":42},{"name":"b","type":"string"}]}
```

This may also imply a bug in the fingerprinting/canonicalization algorithm, since changing the schema to have `"name":"test"` will change the generated fingerprint. I haven't been bitten by this however!

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by brennanvincent**
Certain Avro implementations choke on fullnames with the null namespace (e.g. `".foo"`).

This seems to (maybe?) match the spec, but the spec is confusing and, under the interpretation that it does indeed prohibit leading dots in fullnames, makes it impossible to serialize a large class of logically valid schemas (#14333).

Thus, pending upstream resolution of the questions about the spec, we do _not_ avoid emitting such fullnames in the case where they're indispensable for expressing the schema. But in this PR, we do avoid emitting fullnames _whenever possible_, making a wider class of our schemas interpretable by downstream tools.

### Motivation

Fixes #14333.

### Tips for reviewer

`test_schema_fingerprint` now matches Python:

```
>>> import avro.schema
>>> schema = avro.schema.parse(open("test_schema.json", "rb").read())
>>> import hashlib
>>> hashlib.sha256(schema.canonical_form.encode()).hexdigest()
'c4d97949770866dec733ae7afa3046757e901d0cfea32eb92a8faeadcc4de153'
```

However, I think PCF might still be broken, since it will not fully-qualify anything that doesn't need to be (which seems to violate the spec). I'll look into that as a follow-up.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - Avro schema namespaces are now encoded only when necessary

### CODE REVIEW

**Comment by benesch** [comment]
This idea seems great to me.

File: src/avro/src/schema.rs [hunk 1]
```diff
@@ -1438,6 +1454,11 @@ impl<'a> SchemaNodeOrNamed<'a> {
             top: piece,
         }
     }
+
+    pub fn ns(self) -> Option<&'a str> {
```
Comment (benesch) [inline src/avro/src/schema.rs:1458]: This is spelled out as "namespace" in other public APIs, no?

File: src/avro/src/schema.rs [hunk 2]
```diff
@@ -2839,13 +2876,8 @@ mod tests {
 
         let schema = Schema::from_str(raw_schema).unwrap();
         assert_eq!(
-            "5ecb2d1f0eaa647d409e6adbd5d70cd274d85802aa9167f5fe3b73ba70b32c76",
+            "c4d97949770866dec733ae7afa3046757e901d0cfea32eb92a8faeadcc4de153",
```
Comment (benesch) [inline src/avro/src/schema.rs:2879]: Is this not expected to be stable across different Avro libraries? If yes, I guess the old code was just computing the wrong fingerprint?

**Review by benesch — APPROVED** [review]

File: src/avro/src/schema.rs (hunk 2 above)
Comment (brennanvincent) [reply src/avro/src/schema.rs:2879]: It was computing something which... I interpret to be correct, but the Python avro library's canonical_form doesn't match. Now they do match, which is probably better.

It's still busted, though, on more complicated schemas, (it doesn't actually fill in the fullnames for things that didn't have a `namespace` entry but was getting them from the enclosing schema), but fixing that can be a follow-up.

Comment (brennanvincent) [reply src/avro/src/schema.rs:2879]: (It's no _worse_ than it was before, which is why I want to fix it in a separate PR)

Comment (benesch) [reply src/avro/src/schema.rs:2879]: Yeah, moving in the right direction sounds great!

**Pull request merged**